File: leetcode/java/49. Group Anagrams.java
import java.util.*;
import java.io.*;

class Main {
      
    public static PrintStream ps = System.out;

    public static List<List<String>> groupAnagrams(String[] strs) {
        ArrayList<List<String>> result = new ArrayList<List<String>>();
        HashMap<String, Integer> helper = new HashMap<String, Integer>();
        for (int i = 0; i < strs.length; i ++) {
            char[] tmpChar = strs[i].toCharArray();
            Arrays.sort(tmpChar);
            String tmp = new String(tmpChar);
            if (helper.containsKey(tmp)) {
                result.get(helper.get(tmp)).add(strs[i]);
                continue;
            }
            List<String> curr = new LinkedList<String>();
            curr.add(strs[i]);
            result.add(curr);
            helper.put(tmp, result.size() - 1);
        }
        return result;     
    }

    public static void main(String ...s){
       String[] dataSet = new String[] {"eat","tea","tan","ate","nat","bat"};
       for(List<String> c:groupAnagrams(dataSet)){
         c.stream().forEach(ps::println);
         ps.println("###############");
       }
    }

}

File: leetcode/kotlin/49. Group Anagrams.kt
import java.util.*
import java.io.*

internal object Main {
    var ps: PrintStream = System.out
    fun groupAnagrams(strs: Array<String>): List<List<String>> {
        val result: ArrayList<List<String>> = ArrayList<List<String>>()
        val helper: HashMap<String, Integer> = HashMap<String, Integer>()
        for (i in strs.indices) {
            val tmpChar: CharArray = strs[i].toCharArray()
            Arrays.sort(tmpChar)
            val tmp = String(tmpChar)
            if (helper.containsKey(tmp)) {
                result.get(helper.get(tmp)).add(strs[i])
                continue
            }
            val curr: List<String> = LinkedList<String>()
            curr.add(strs[i])
            result.add(curr)
            helper.put(tmp, result.size() - 1)
        }
        return result
    }

    fun main(vararg s: String?) {
        val dataSet = arrayOf("eat", "tea", "tan", "ate", "nat", "bat")
        for (c in groupAnagrams(dataSet)) {
            c.stream().forEach(ps::println)
            ps.println("###############")
        }
    }
}